File: Lecture1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

string validateCreditCardNumber(long long const card_number);
long long power(int base, int exponent);

int main(void)
{
    long long number = get_long_long("Number: ");
    printf("%s\n", validateCreditCardNumber(number));
}

string validateCreditCardNumber(long long const card_number)
{
    // Luhn's validation
    int checksum = 0;
    int card_length = 0;

    for (long long end_number_test = card_number; end_number_test != 0; end_number_test /= 10)
    {
        int digit = end_number_test % 10;
        if (card_length % 2 == 1)
        {
            digit *= 2;
        }
        checksum += digit / 10 + digit % 10;
        card_length++;
    }
    if (checksum % 10 != 0)
    {
        return "INVALID";
    }

    int first_two_digits = card_number / power(10, card_length - 2);
    int first_digit = first_two_digits / 10;

    if ((first_two_digits == 34 || first_two_digits == 37) && card_length == 15)
    {
        return "AMEX";
    }
    if (51 <= first_two_digits && first_two_digits <= 55 && card_length == 16)
    {
        return "MASTERCARD";
    }
    if (first_digit == 4 && (card_length == 13 || card_length == 16))
    {
        return "VISA";
    }
    return "INVALID";
}

long long power(int const base, int const exponent)
{
    if (exponent == 1)
    {
        return base;
    }
    return base * power(base, exponent - 1);
}

File: lecture6/sentimental-credit/credit.py
def main():
    card_number = int(input("Number: "))

    str_card_number = str(card_number)
    card_length = len(str_card_number)

    if not is_valid_credit_card_number(str_card_number):
        print("INVALID")
        return
    if is_amex(str_card_number, card_length):
        print("AMEX")
        return
    if is_mastercard(str_card_number, card_length):
        print("MASTERCARD")
        return
    if is_visa(str_card_number, card_length):
        print("VISA")
        return

    print("INVALID")


def is_valid_credit_card_number(str_card_number: str) -> bool:
    check_sum = 0
    for index, digit in enumerate(str_card_number[::-1]):
        digit = int(digit)
        if int(index % 2) == 1:
            digit *= 2
        check_sum += int(digit / 10 + digit % 10)
    if check_sum % 10 != 0:
        return False
    return True


def is_amex(str_card_number: str, card_length: int) -> bool:
    first_two_digits = str_card_number[:2]
    if first_two_digits in ("34", "37") and card_length == 15:
        return True
    return False


def is_mastercard(str_card_number: str, card_length: int) -> bool:
    first_two_digits = str_card_number[:2]
    if int(first_two_digits) in range(51, 56) and card_length == 16:
        return True
    return False


def is_visa(str_card_number: str, card_length: int) -> bool:
    first_digit = str_card_number[0]
    if first_digit == "4" and (card_length in (13, 16)):
        return True
    return False


if __name__ == "__main__":
    main()
